core/src/main/java/io/grpc/transport/MessageFramer.java

```diff
@@ -270,26 +317,53 @@ public void write(byte[] b, int off, int len) {
   }
 
   /**
-   * Implements the same general contract of DeferredInputStream, although is unable to extend it.
+   * Produce a collection of {@link WritableBuffer} instances from the data written to an
+   * {@link OutputStream}.
    */
-  private static class DeferredByteArrayInputStream extends ByteArrayInputStream {
-    public DeferredByteArrayInputStream(byte[] buf, int offset, int length) {
-      super(buf, offset, length);
+  private class BufferChainOutputStream extends OutputStream {
+
+    private final byte[] singleByte = new byte[1];
+    private List<WritableBuffer> bufferList;
+    private WritableBuffer current;
+
+    private BufferChainOutputStream() {
+      bufferList = new ArrayList<WritableBuffer>();
     }
 
-    public int flushTo(OutputStream os) throws IOException {
-      os.write(buf, pos, count - pos);
-      return count - pos;
+    @Override
+    public void write(int b) throws IOException {
+      singleByte[0] = (byte) b;
+      write(singleByte, 0, 1);
     }
-  }
 
-  private static class DirectAccessByteArrayOutputStream extends ByteArrayOutputStream {
-    public byte[] getBuf() {
-      return buf;
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+      if (current == null) {
+        // Request at least one byte, the allocator will typically allocate more than that.
+        current = bufferAllocator.allocate(1);
```

Comment:
The allocator will really take care of making sure we don't get anything uselessly small. Using available isn't really going to help with that unless it returns a value greater than the min buffer size in the allocator which would be odd as that would indicate that the stream had a more accurate idea about its real length in the first place.

Using available if its > 0